data/src/dcc.rs

```diff
@@ -66,12 +66,37 @@ impl Send {
         }
     }
 
-    fn decode<'a>(mut args: impl Iterator<Item = &'a str>) -> Option<Self> {
-        let filename = args.next()?.to_string();
-        let host = args.next().and_then(decode_host)?;
-        let port = NonZeroU16::new(args.next()?.parse().ok()?);
-        let size = args.next()?.parse().ok()?;
-        let token = args.next();
+    fn decode<'a>(args: impl Iterator<Item = &'a str>) -> Option<Self> {
+        let mut args: Vec<&str> = args.collect();
+        args.reverse();
+
+        // if token doesn't exist
+        let mut token = None;
+        let mut size = args[0];
+        let mut port = args[1];
+        let mut host = args[2];
+        let mut filename = args.iter().skip(3);
+
+        // If token exists, port == 0
+        // args[1] == port, if token doesn't exists.
+        // args[2] == port, if token exists.
+        if args[1].parse::<u16>() == Ok(0) || args[2].parse::<u16>() == Ok(0) {
```

Comment:
Slicing is unsafe as we haven't validated args >= 3 so these can panic / crash the application

Reply:
after changing the logic to use iterators, this doesn't apply anymore.